pkg/autoupdate/findnew.go

```diff
@@ -0,0 +1,177 @@
+package autoupdate
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/kolide/launcher/pkg/contexts/ctxlog"
+	"github.com/pkg/errors"
+)
+
+// This suffix is added to the binary path to find the updates
+const updateDirSuffix = "-updates"
+
+type newestSettings struct {
+	deleteOld bool
+}
+
+type newestOption func(*newestSettings)
+
+func DeleteOldUpdates() newestOption {
+	return func(no *newestSettings) {
+		no.deleteOld = true
+	}
+}
+
+// FindNewestSelf invokes `FindNewest` with the running binary path,
+// as determined by os.Executable. However, if the current running
+// version is the same as the newest on disk, it will return empty string.
+func FindNewestSelf(ctx context.Context, opts ...newestOption) (string, error) {
+	logger := log.With(ctxlog.FromContext(ctx), "caller", log.DefaultCaller)
+
+	exPath, err := os.Executable()
+	if err != nil {
+		return "", errors.Wrap(err, "determine running executable path")
+	}
+
+	if exPath == "" {
+		return "", errors.New("can't find newest empty string")
+	}
+
+	newest := FindNewest(ctx, exPath, opts...)
+
+	if newest == "" {
+		return "", nil
+	}
+
+	if exPath == newest {
+		return "", nil
+	}
+
+	level.Debug(logger).Log(
+		"msg", "found an update",
+		"newest", newest,
+		"exPath", exPath,
+	)
+
+	return newest, nil
+}
+
+// FindNewest takes the full path to a binary, and returns the newest
+// update on disk. If there are no updates on disk, it returns the
+// original path. It will return the same fullBinaryPath if that is
+// the newest version.
+func FindNewest(ctx context.Context, fullBinaryPath string, opts ...newestOption) string {
+	logger := log.With(ctxlog.FromContext(ctx), "caller", log.DefaultCaller)
+
+	if fullBinaryPath == "" {
+		level.Debug(logger).Log("msg", "called with empty string")
+		return ""
+	}
+
+	logger = log.With(logger, "fullBinaryPath", fullBinaryPath)
+
+	newestSettings := &newestSettings{}
+	for _, opt := range opts {
+		opt(newestSettings)
+	}
+
+	updateDir := getUpdateDir(fullBinaryPath)
+	binaryName := filepath.Base(fullBinaryPath)
+
+	fileGlob := filepath.Join(updateDir, "*", binaryName)
+
+	possibleUpdates, err := filepath.Glob(fileGlob)
+	if err != nil {
```

Comment:
`Glob(..)` returns nil if no matches are found, should we return early or if there are no matches found _and_ the err is nil? Not sure what that case means (probably that launcher is running but it's on-disk binary is moved or missing..?)
Reading the docs, I think we should probably handle this case, since errors are only thrown for invalid input:
`Glob ignores file system errors such as I/O errors reading directories. The only possible returned error is ErrBadPattern, when pattern is malformed.`

Also, further on in the code it looks like this will actually be handled fine, but I think I'd want some logs for this if I was debugging it.

Reply:
I don't think an early return makes sense. If there are no matches, then we return either `fullBinaryPath` or `""` depending. I don't think duplicating that logic makes sense.

I did, however, add a comment.